{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding files and data using Data Connect\n",
    "\n",
    "#### Learning Objectives\n",
    "Workshop attendees will learn how use the GA4GH Data Connect Service.  \n",
    "\n",
    "What will participants do as part of the exercise?\n",
    "\n",
    " - Understanding how to query data via Data Connect\n",
    " - Use Data Connect to find files that can be accessed via DRS\n",
    " - Learn how to obtain and use data deacriptions (schema)\n",
    " - Discover the meaning of codes used in data\n",
    " \n",
    "\n",
    " #### Icons in this Guide\n",
    "\n",
    " üñê A hands-on section where you will code something or interact with the server\n",
    " \n",
    "### Query files\n",
    "The approach taken below is using mapping is available through subject and specimen data available through the Data Connect API. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fasp.search import DataConnectClient\n",
    "\n",
    "# Step 1 - Discovery\n",
    "# query for relevant DRS objects\n",
    "searchClient = DataConnectClient('http://localhost:8089/', debug=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n",
      "____Page9_______________\n",
      "____Page10_______________\n",
      "____Page11_______________\n",
      "____Page12_______________\n",
      "____Page13_______________\n",
      "____Page14_______________\n",
      "____Page15_______________\n",
      "____Page16_______________\n",
      "____Page17_______________\n",
      "____Page18_______________\n",
      "____Page19_______________\n",
      "____Page20_______________\n",
      "____Page21_______________\n",
      "____Page22_______________\n",
      "____Page23_______________\n",
      "____Page24_______________\n",
      "____Page25_______________\n",
      "____Page26_______________\n",
      "____Page27_______________\n",
      "____Page28_______________\n",
      "____Page29_______________\n",
      "____Page30_______________\n",
      "____Page31_______________\n",
      "____Page32_______________\n",
      "____Page33_______________\n",
      "____Page34_______________\n",
      "____Page35_______________\n",
      "____Page36_______________\n",
      "____Page37_______________\n",
      "____Page38_______________\n",
      "____Page39_______________\n",
      "____Page40_______________\n",
      "____Page41_______________\n",
      "____Page42_______________\n",
      "____Page43_______________\n",
      "____Page44_______________\n",
      "____Page45_______________\n",
      "____Page46_______________\n",
      "____Page47_______________\n",
      "____Page48_______________\n",
      "____Page49_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>bam_drs_id</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA18948</td>\n",
       "      <td>fb1cfb04d3ef99d07c21f9dbf87ccc68</td>\n",
       "      <td>SRR1601121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA18945</td>\n",
       "      <td>9327fb44eb81b49a41e38c8d86eb3b3a</td>\n",
       "      <td>SRR1601115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA18943</td>\n",
       "      <td>9f38253b281c7e9c99e4bdbececd8e2f</td>\n",
       "      <td>SRR1606910</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name                        bam_drs_id         acc\n",
       "0     NA18948  fb1cfb04d3ef99d07c21f9dbf87ccc68  SRR1601121\n",
       "1     NA18945  9327fb44eb81b49a41e38c8d86eb3b3a  SRR1601115\n",
       "2     NA18943  9f38253b281c7e9c99e4bdbececd8e2f  SRR1606910"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = '''SELECT f.sample_name, drs_id bam_drs_id, acc\n",
    "FROM tutorial.onek_genomes.ssd_drs s \n",
    "join tutorial.onek_genomes.sra_drs_files f on f.sample_name = s.su_submitter_id \n",
    "where filetype = 'bam' and mapped = 'mapped' \n",
    "and sequencing_type ='exome' and  population = 'JPT' LIMIT 3'''\n",
    "\n",
    "resultRows = searchClient.run_query(query, returnType='dataframe')\n",
    "resultRows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use DRS to obtain file details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tabletutorial.onek_genomes.sra_drs_files_\n",
      "{\n",
      "   \"name\": \"tutorial.onek_genomes.sra_drs_files\",\n",
      "   \"description\": \"Automatically generated schema\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"http://localhost:8089/table/tutorial.onek_genomes.sra_drs_files/info\",\n",
      "      \"description\": \"Automatically generated schema\",\n",
      "      \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n",
      "      \"properties\": {\n",
      "         \"acc\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"filename\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"drs_id\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"filetype\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"sample_name\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"mapped\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         },\n",
      "         \"sequencing_type\": {\n",
      "            \"format\": \"varchar\",\n",
      "            \"type\": \"string\",\n",
      "            \"$comment\": \"varchar\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x12ff64430>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "searchClient.list_table_info('tutorial.onek_genomes.sra_drs_files', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tabletutorial.onek_genomes.ssd_drs_\n",
      "{\n",
      "   \"name\": \"tutorial.onek_genomes.ssd_drs\",\n",
      "   \"description\": \"Thousand genomes sample metadata\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"t5\",\n",
      "      \"description\": \"Thousand genomes sample metadata\",\n",
      "      \"properties\": {\n",
      "         \"population\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"v4\",\n",
      "            \"description\": \"Population list\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"ACB\",\n",
      "                  \"title\": \"African Caribbean in Barbados\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"ASW\",\n",
      "                  \"title\": \"African Ancestry in Southwest US\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"BEB\",\n",
      "                  \"title\": \"Bengali in Bangladesh\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CDX\",\n",
      "                  \"title\": \"Chinese Dai in Xishuangbanna, China\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CEU\",\n",
      "                  \"title\": \"Utah residents (CEPH) with Northern and Western European ancestry \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CHB\",\n",
      "                  \"title\": \"Han Chinese in Beijing, China\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CHS\",\n",
      "                  \"title\": \"Han Chinese South\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CLM\",\n",
      "                  \"title\": \"Colombian in Medellin, Colombia\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"ESN\",\n",
      "                  \"title\": \"Esan in Nigeria\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"FIN\",\n",
      "                  \"title\": \"Finnish in Finland \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"GBR\",\n",
      "                  \"title\": \"British in England and Scotland \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"GIH\",\n",
      "                  \"title\": \"Gujarati Indian in Houston, TX\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"GWD\",\n",
      "                  \"title\": \"Gambian in Western Division, The Gambia \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"IBS\",\n",
      "                  \"title\": \"Iberian populations in Spain \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"ITU\",\n",
      "                  \"title\": \"Indian Telugu in the UK\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"JPT\",\n",
      "                  \"title\": \"Japanese in Tokyo, Japan\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"KHV\",\n",
      "                  \"title\": \"Kinh in Ho Chi Minh City, Vietnam\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"LWK\",\n",
      "                  \"title\": \"Luhya in Webuye, Kenya\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"MSL\",\n",
      "                  \"title\": \"Mende in Sierra Leone\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"MXL\",\n",
      "                  \"title\": \"Mexican Ancestry in Los Angeles, California\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"PEL\",\n",
      "                  \"title\": \"Peruvian in Lima, Peru\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"PJL\",\n",
      "                  \"title\": \"Punjabi in Lahore, Pakistan\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"PUR\",\n",
      "                  \"title\": \"Puerto Rican in Puerto Rico\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"STU\",\n",
      "                  \"title\": \"Sri Lankan Tamil in the UK\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"TSI\",\n",
      "                  \"title\": \"Toscani in Italia \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"YRI\",\n",
      "                  \"title\": \"Yoruba in Ibadan, Nigeria\"\n",
      "               }\n",
      "            ]\n",
      "         },\n",
      "         \"race\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"v1\",\n",
      "            \"description\": \"Race no list\"\n",
      "         },\n",
      "         \"ethnicity\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"v2\",\n",
      "            \"description\": \"Ethnicity no list\"\n",
      "         },\n",
      "         \"weight\": {\n",
      "            \"type\": \"number\",\n",
      "            \"$id\": \"v3\",\n",
      "            \"description\": \"weight in kg\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<fasp.search.data_connect_client.SearchSchema at 0x12ff64dc0>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "searchClient.list_table_info('tutorial.onek_genomes.ssd_drs', verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the information above about the tables we can modify the query \n",
    "\n",
    "#superpopulation_name = \"European Ancestry\" and population_code = \"FIN\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n",
      "____Page9_______________\n",
      "____Page10_______________\n",
      "____Page11_______________\n",
      "____Page12_______________\n",
      "____Page13_______________\n",
      "____Page14_______________\n",
      "____Page15_______________\n",
      "____Page16_______________\n",
      "____Page17_______________\n",
      "____Page18_______________\n",
      "____Page19_______________\n",
      "____Page20_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = '''SELECT weight\n",
    "FROM tutorial.onek_genomes.ssd_drs s \n",
    "where weight is not null\n",
    "limit 10 '''\n",
    "\n",
    "resultRows = searchClient.run_query(query, returnType='dataframe')\n",
    "resultRows\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n",
      "____Page9_______________\n",
      "____Page10_______________\n",
      "____Page11_______________\n",
      "____Page12_______________\n",
      "____Page13_______________\n",
      "____Page14_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>pop_cnt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACB</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASW</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BEB</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDX</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CEU</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHB</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CLM</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ESN</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FIN</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GBR</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GIH</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GWD</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>IBS</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ITU</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>JPT</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>KHV</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>LWK</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MSL</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MXL</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PEL</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PJL</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PUR</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>STU</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TSI</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>YRI</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   population  pop_cnt\n",
       "0         ACB       96\n",
       "1         ASW       61\n",
       "2         BEB       86\n",
       "3         CDX       93\n",
       "4         CEU       99\n",
       "5         CHB      103\n",
       "6         CHS      105\n",
       "7         CLM       94\n",
       "8         ESN       99\n",
       "9         FIN       99\n",
       "10        GBR       91\n",
       "11        GIH      103\n",
       "12        GWD      113\n",
       "13        IBS      107\n",
       "14        ITU      102\n",
       "15        JPT      104\n",
       "16        KHV       99\n",
       "17        LWK       99\n",
       "18        MSL       85\n",
       "19        MXL       64\n",
       "20        PEL       85\n",
       "21        PJL       96\n",
       "22        PUR      104\n",
       "23        STU      102\n",
       "24        TSI      107\n",
       "25        YRI      108"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = '''SELECT population, count(*) pop_cnt\n",
    "FROM tutorial.onek_genomes.ssd_drs s \n",
    "group by population\n",
    "order by population'''\n",
    "\n",
    "resultRows = searchClient.run_query(query, returnType='dataframe')\n",
    "resultRows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT f.sample_name, drs_id bam_drs_id, acc, filename\n",
      "FROM thousand_genomes.onek_genomes.ssd_drs s \n",
      "join thousand_genomes.onek_genomes.sra_drs_files f on f.sample_name = s.su_submitter_id \n",
      "\n",
      "where filetype = 'bam' and mapped = 'mapped' \n",
      "and sequencing_type ='exome' and  population = 'PUR' \n",
      "and  filename like '%HG00740%'\n",
      "\n",
      "LIMIT 3\n"
     ]
    }
   ],
   "source": [
    "query = '''SELECT f.sample_name, drs_id bam_drs_id, acc, filename\n",
    "FROM thousand_genomes.onek_genomes.ssd_drs s \n",
    "join thousand_genomes.onek_genomes.sra_drs_files f on f.sample_name = s.su_submitter_id \n",
    "\n",
    "where filetype = 'bam' and mapped = 'mapped' \n",
    "and sequencing_type ='exome' and  population = 'PUR' \n",
    "and  filename like '%HG00740%'\n",
    "\n",
    "LIMIT 3'''\n",
    "\n",
    "print(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT f.sample_name, drs_id bam_drs_id, acc, filename\n",
      "FROM thousand_genomes.onek_genomes.ssd_drs s \n",
      "join thousand_genomes.onek_genomes.sra_drs_files f on f.sample_name = s.su_submitter_id \n",
      "\n",
      "where filename like 'HG00740%'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "query = '''SELECT f.sample_name, drs_id bam_drs_id, acc, filename\n",
    "FROM thousand_genomes.onek_genomes.ssd_drs s \n",
    "join thousand_genomes.onek_genomes.sra_drs_files f on f.sample_name = s.su_submitter_id \n",
    "\n",
    "where filename like 'HG00740%'\n",
    "'''\n",
    "print(query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The method of calling the Data Connect client above returns a dataframe. This is convenient for many purposes, including listing the results as above. The default return type from the runQuery gives a list of lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n",
      "____Page8_______________\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[['HG00740',\n",
       "  '8b8388fdc4bef15213df69a5a892ba47',\n",
       "  'SRR1596984',\n",
       "  'HG00740.unmapped.ILLUMINA.bwa.PUR.low_coverage.20120522.bam'],\n",
       " ['HG00740',\n",
       "  '2a5a92225137d5abc65a2b9fadd25264',\n",
       "  'SRR1596984',\n",
       "  'HG00740.mapped.ILLUMINA.bwa.PUR.low_coverage.20120522.bam'],\n",
       " ['HG00740',\n",
       "  'e78af56200c0c6ee4c08ab043b1f71bb',\n",
       "  'SRR1596983',\n",
       "  'HG00740.unmapped.ILLUMINA.bwa.PUR.exome.20120522.bam'],\n",
       " ['HG00740',\n",
       "  '61eded72b0d2dfc908fee2201792dbac',\n",
       "  'SRR1596983',\n",
       "  'HG00740.mapped.ILLUMINA.bwa.PUR.exome.20120522.bam']]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = searchClient.run_query(query)\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following shows how the SRA DRS server can be used to determine where the files can be obtained from. The following shows this for the first DRS id from the query results. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8b8388fdc4bef15213df69a5a892ba47\n",
      "https://locate.be-md.ncbi.nlm.nih.gov/ga4gh/drs/v1/objects/8b8388fdc4bef15213df69a5a892ba47\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'access_methods': [{'access_id': '0814ace3869680a709dadeb570293cd0c80f2d2e9f0f49a96010e5315db37f6a',\n",
       "   'region': 'gs.US',\n",
       "   'type': 'https'},\n",
       "  {'access_id': '5b998d00d9c39e3e476d99855f9f53e069294435568029f89caf5252c87b991f',\n",
       "   'type': 'https'},\n",
       "  {'access_id': 'f27a3b7058835391094fd19ef63c4062a8f219c72016906a131324048b1dfae1',\n",
       "   'region': 's3.us-east-1',\n",
       "   'type': 'https'}],\n",
       " 'checksums': [{'checksum': '8b8388fdc4bef15213df69a5a892ba47',\n",
       "   'type': 'md5'}],\n",
       " 'created_time': '2012-11-16T02:33:10Z',\n",
       " 'id': '8b8388fdc4bef15213df69a5a892ba47',\n",
       " 'name': 'HG00740.unmapped.ILLUMINA.bwa.PUR.low_coverage.20120522.bam',\n",
       " 'self_url': 'drs://locate.be-md.ncbi.nlm.nih.gov/8b8388fdc4bef15213df69a5a892ba47',\n",
       " 'size': 362537865}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from fasp.loc import DRSClient\n",
    "\n",
    "#drsClient = DRSMetaResolver()\n",
    "drsClient = DRSClient('https://locate.be-md.ncbi.nlm.nih.gov', public=True, debug=True)\n",
    "test_id = results[0][1]\n",
    "print(test_id)\n",
    "objInfo = drsClient.get_object(test_id)\n",
    "objInfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A second DRS call can be used to obtain a url to access the file from one of the above locations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "access_id:0814ace3869680a709dadeb570293cd0c80f2d2e9f0f49a96010e5315db37f6a\n",
      "https://locate.be-md.ncbi.nlm.nih.gov/ga4gh/drs/v1/objects/8b8388fdc4bef15213df69a5a892ba47/access/0814ace3869680a709dadeb570293cd0c80f2d2e9f0f49a96010e5315db37f6a\n",
      "<Response [200]>\n",
      "url:https://storage.googleapis.com/genomics-public-data/ftp-trace.ncbi.nih.gov/1000genomes/ftp/phase3/data/HG00740/alignment/HG00740.unmapped.ILLUMINA.bwa.PUR.low_coverage.20120522.bam\n"
     ]
    }
   ],
   "source": [
    "access_id = objInfo['access_methods'][0]['access_id']\n",
    "print('access_id:{}'.format(access_id))\n",
    "url = drsClient.get_access_url(test_id, access_id=access_id)\n",
    "print('url:{}'.format(url))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
